iocore/net/quic/QUICTLS.cc

```diff
@@ -202,6 +380,119 @@ QUICTLS::abort_handshake()
   return;
 }
 
+void
+QUICTLS::set_ready_for_write()
+{
+  this->_should_flush = true;
+}
+
+void
+QUICTLS::on_handshake_data_generated(QUICEncryptionLevel level, const uint8_t *data, size_t len)
+{
+  int index      = static_cast<int>(level);
+  int next_index = index + 1;
+
+  size_t offset            = this->_out.offsets[next_index];
+  size_t next_level_offset = offset + len;
+
+  memcpy(this->_out.buf + offset, data, len);
+
+  for (int i = next_index; i < 5; ++i) {
+    this->_out.offsets[i] = next_level_offset;
+  }
+}
+
+void
+QUICTLS::on_tls_alert(uint8_t alert)
+{
+  this->_has_crypto_error = true;
+  this->_crypto_error     = QUICTLS::convert_to_quic_trans_error_code(alert);
+}
+
+bool
+QUICTLS::has_crypto_error() const
+{
+  return this->_has_crypto_error;
+}
+
+uint64_t
+QUICTLS::crypto_error() const
+{
+  return this->_crypto_error;
+}
+
+int
+QUICTLS::_handshake(QUICHandshakeMsgs **out, const QUICHandshakeMsgs *in)
+{
+  ink_assert(this->_ssl != nullptr);
+  if (this->_state == HandshakeState::ABORTED) {
+    return 0;
+  }
+
+  int err = SSL_ERROR_NONE;
+  ERR_clear_error();
+  int ret = 0;
+
+  SSL_set_msg_callback(this->_ssl, QUICTLS::_msg_cb);
+
+  this->_out.offsets[0] = 0;
+  this->_out.offsets[1] = 0;
+  this->_out.offsets[2] = 0;
+  this->_out.offsets[3] = 0;
+  this->_out.offsets[4] = 0;
+
+  if (in) {
+    this->_pass_quic_data_to_ssl_impl(*in);
+  }
+
+  if (this->_netvc_context == NET_VCONNECTION_IN) {
+    if (!this->_early_data_processed) {
+      if (auto ret = this->_read_early_data(); ret == 0) {
+        this->_early_data_processed = true;
+      } else if (ret < 0) {
+        (*out)->error_code = static_cast<uint16_t>(QUICTransErrorCode::PROTOCOL_VIOLATION);
```

Comment:
can *out be nullptr ?

Reply:
This line was not needed. I pushed a fix for this.